plugin/vcs/gitlab/gitlab.go

```diff
@@ -479,84 +481,111 @@ func getRoleAndMappedRole(accessLevel int32) (gitLabRole ProjectRole, bytebaseRo
-	)
-	if err != nil {
-		return nil, errors.Wrap(err, "GET")
-	}
-
-	if code == http.StatusNotFound {
-		return nil, common.Errorf(common.NotFound, fmt.Errorf("failed to fetch repository members from GitLab instance %s", instanceURL))
-	} else if code >= 300 {
-		return nil, fmt.Errorf("failed to read repository members from GitLab instance %s, status code: %d",
-			instanceURL,
-			code,
+	// Preallocate the slice to avoid multiple allocations when appending the first
+	// 100 results to the slice.
+	allMembers := make([]RepositoryMember, 0, 100)
+
+	// The "state" filter only available in GitLab Premium self-managed, GitLab
+	// Premium SaaS, and higher tiers, but worth a try for less abandoned results.
+	nextURL := fmt.Sprintf("%s/projects/%s/members/all?state=active&per_page=100", p.APIURL(instanceURL), repositoryID)
+	for nextURL != "" {
+		code, header, body, err := oauth.Get(
+			ctx,
+			p.client,
+			nextURL,
+			&oauthCtx.AccessToken,
+			tokenRefresher(
+				instanceURL,
+				oauthContext{
+					ClientID:     oauthCtx.ClientID,
+					ClientSecret: oauthCtx.ClientSecret,
+					RefreshToken: oauthCtx.RefreshToken,
+				},
+				oauthCtx.Refresher,
+			),
 		)
-	}
+		if err != nil {
+			return nil, errors.Wrapf(err, "GET %s", nextURL)
+		}
 
-	var gitLabrepositoryMember []gitLabRepositoryMember
-	if err := json.Unmarshal([]byte(body), &gitLabrepositoryMember); err != nil {
-		return nil, err
+		if code == http.StatusNotFound {
+			return nil, common.Errorf(common.NotFound, fmt.Errorf("failed to fetch repository members from URL %s", nextURL))
+		} else if code >= 300 {
+			return nil, fmt.Errorf("failed to read repository members from URL %s, status code: %d, body: %s",
+				nextURL,
+				code,
+				body,
+			)
+		}
+
+		var members []RepositoryMember
+		if err := json.Unmarshal([]byte(body), &members); err != nil {
+			return nil, errors.Wrap(err, "unmarshal body")
+		}
+		allMembers = append(allMembers, members...)
+
+		if l := link.Parse(header.Get("Link"))["next"]; l != nil {
+			nextURL = l.URI
+		} else {
+			nextURL = ""
+		}
 	}
 
-	// we only return active member (both state and membership_state is active)
 	var emptyEmailUserIDList []string
-	var activeRepositoryMemberList []*vcs.RepositoryMember
-	for _, gitLabMember := range gitLabrepositoryMember {
-		if gitLabMember.State == vcs.StateActive {
-			// The email field will only be returned if the caller credential is associated with a GitLab admin account.
-			// And since most callers are not GitLab admins, thus we fetch public email
+	var activeMembers []*vcs.RepositoryMember
+	for _, m := range allMembers {
+		// We only want active member (both state and membership_state)
+		if m.State != vcs.StateActive || m.MembershipState != vcs.StateActive {
+			continue
+		}
+
+		// The email field is only returned if the caller credential is associated with
+		// a GitLab admin account. We'll try getting the email by fetching the info of
+		// individual users.
+		if m.Email == "" {
```

Comment:
Save API calls if the credentials happens to be from the admin, if not, still doesn't hurt.